File: back-end/src/main/java/pl/straburzynski/campfiresongs/externalapi/service/parser/SpiewnikWywrotaParserService.java
package pl.straburzynski.campfiresongs.externalapi.service.parser;

import lombok.extern.slf4j.Slf4j;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import pl.straburzynski.campfiresongs.externalapi.exception.ExternalApiException;
import pl.straburzynski.campfiresongs.externalapi.model.ExternalApiResponse;
import pl.straburzynski.campfiresongs.externalapi.model.ExternalApiSong;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static pl.straburzynski.campfiresongs.externalapi.model.ExternalApiSource.SPIEWNIK_WYWROTA;

@Slf4j
public class SpiewnikWywrotaParserService {

    private final static String DEFAULT_TAG_NAME = "code";
    private final static String ANNOTATION_CLASS = "class=\"an\"";
    private final static String CONTENT = "interpretation-content";
    private final static String NEW_LINE_CHARACTER = "\n";
    private final static String ANNOTATION_TAG_START = "<span class=\"annotated-lyrics\">";
    private final static String ANNOTATION_TAG_END = "</span>";
    private final static String MUTED_TAG_START = "<span class=\"text-muted\">";
    private final static String HTML_BREAK_LINE = "<br>";
    private final static String HTML_HARD_SPACE = "&nbsp;";

    public static String getRegexPatternForTag() {
        return "<" + DEFAULT_TAG_NAME + ".*?>(.+?)</" + DEFAULT_TAG_NAME + ">";
    }

    public static String convertTagToChord(String code, boolean isAnnotation) {
        return convertTagToChord(DEFAULT_TAG_NAME, code, isAnnotation);
    }

    public static String convertTagToChord(String tagName, String tagWithContent, boolean isAnnotation) {
        Element code = Jsoup.parseBodyFragment(tagWithContent).getElementsByTag(tagName).first();
        if (code == null) {
            log.debug("Didn't find any value in tag, {}", tagWithContent);
            return "";
        } else {
            String chordName = code.attr("data-chord") + code.attr("data-suffix");
            return "|" + (isAnnotation ? "@" : "#") + chordName + "|";
        }
    }

    public static String convertRawLineToChordFormatted(String line) {
        boolean isAnnotation = line.contains(ANNOTATION_CLASS);
        Pattern pattern = Pattern.compile(getRegexPatternForTag());
        Matcher matcher = pattern.matcher(line);
        while (matcher.find()) {
            line = line.replace(matcher.group(0), convertTagToChord(matcher.group(0), isAnnotation));
        }
        return line;
    }

    public static List<String> convertDocumentToLines(Document document) {
        Element contentElement = document.getElementsByClass(CONTENT).stream().findFirst()
                .orElseThrow(() -> new ExternalApiException("parseDocument"));
        String[] split = contentElement.html().split(NEW_LINE_CHARACTER);
        List<String> lines = Arrays.asList(split);
        return lines.stream().map(line -> line
                .replaceAll(ANNOTATION_TAG_START, "")
                .replaceAll(MUTED_TAG_START, "")
                .replaceAll(ANNOTATION_TAG_END, "")
                .replaceAll(HTML_BREAK_LINE, "")
                .replaceAll(HTML_HARD_SPACE, "")
        ).collect(Collectors.toList());
    }

    public static String parseLyricsFromDocument(Document document) {
        List<String> lines = convertDocumentToLines(document);
        return lines.stream()
                .map(SpiewnikWywrotaParserService::convertRawLineToChordFormatted)
                .collect(Collectors.joining(NEW_LINE_CHARACTER));
    }

    public static List<ExternalApiSong> parseSearchResults(ExternalApiResponse response) {
        if (response != null) {
            if (response.getSongs() != null) {
                return response.getSongs().stream()
                        .map(song -> ExternalApiSong.builder()
                                .artist(song.getArtist())
                                .title(song.getTitle())
                                .url(song.getUrl())
                                .source(SPIEWNIK_WYWROTA)
                                .build()
                        )
                        .collect(Collectors.toList());
            }
        }
        return Collections.emptyList();
    }
}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/externalapi/service/parser/SpiewnikWywrotaParserService.kt
package pl.straburzynski.campfiresongs.externalapi.service.parser

import java.util.logging.Logger
import java.util.regex.Pattern
import org.jsoup.Jsoup.parseBodyFragment
import org.jsoup.nodes.Document
import pl.straburzynski.campfiresongs.externalapi.exception.ExternalApiException
import pl.straburzynski.campfiresongs.externalapi.model.ExternalApiResponse
import pl.straburzynski.campfiresongs.externalapi.model.ExternalApiSong
import pl.straburzynski.campfiresongs.externalapi.model.ExternalApiSource.SPIEWNIK_WYWROTA

class SpiewnikWywrotaParserService {

    companion object {
        private val logger: Logger = Logger.getLogger(this::class.java.name)

        private const val DEFAULT_TAG_NAME: String = "code"
        private const val ANNOTATION_CLASS: String = "class=\"an\""
        private const val CONTENT: String = "interpretation-content"
        private const val NEW_LINE_CHARACTER: String = "\n"
        private const val ANNOTATION_TAG_START: String = "<span class=\"annotated-lyrics\">"
        private const val ANNOTATION_TAG_END: String = "</span>"
        private const val MUTED_TAG_START: String = "<span class=\"text-muted\">"
        private const val HTML_BREAK_LINE: String = "<br>"
        private const val HTML_HARD_SPACE: String = "&nbsp;"

        private val REGEX_PATTERN = Pattern.compile("<$DEFAULT_TAG_NAME.*?>(.+?)</$DEFAULT_TAG_NAME>")

        private fun convertDocumentToLines(document: Document): List<String> {
            val contentElement = document.getElementsByClass(CONTENT).stream().findFirst()
                .orElseThrow { ExternalApiException("parseDocument") }
            val lines = contentElement.html().split(NEW_LINE_CHARACTER)
            return lines.map {
                it
                    .replace(ANNOTATION_TAG_START, "")
                    .replace(MUTED_TAG_START, "")
                    .replace(ANNOTATION_TAG_END, "")
                    .replace(HTML_BREAK_LINE, "")
                    .replace(HTML_HARD_SPACE, "")
            }.toList()
        }

        fun parseLyricsFromDocument(document: Document): String {
            val lines: List<String> = convertDocumentToLines(document)
            return lines
                .map { convertRawLineToChordFormatted(it) }
                .joinToString(NEW_LINE_CHARACTER)
        }

        fun parseSearchResults(response: ExternalApiResponse?): List<ExternalApiSong> =
            if (response?.songs != null) {
                response.songs.map {
                    ExternalApiSong(
                        it.title,
                        it.url,
                        it.artist,
                        SPIEWNIK_WYWROTA
                    )
                }.toList()
            } else emptyList()


        fun convertTagToChord(code: String, isAnnotation: Boolean): String =
            convertTagToChord(DEFAULT_TAG_NAME, code, isAnnotation)

        fun convertTagToChord(tagName: String, tagWithContent: String, isAnnotation: Boolean): String {
            val code = parseBodyFragment(tagWithContent).getElementsByTag(tagName).first()
            return if (code == null) {
                logger.info("Didn't find any value in tag: $tagWithContent")
                ""
            } else {
                val chordName = code.attr("data-chord") + code.attr("data-suffix")
                if (chordName == "") return ""
                "|${if (isAnnotation) "@" else "#"}$chordName|"
            }
        }

        fun convertRawLineToChordFormatted(line: String): String {
            var convertedLine = line
            val isAnnotation = convertedLine.contains(ANNOTATION_CLASS)
            val matcher = REGEX_PATTERN.matcher(convertedLine)
            while (matcher.find()) {
                convertedLine = convertedLine.replace(
                    matcher.group(0), convertTagToChord(matcher.group(0), isAnnotation)
                )
            }
            return convertedLine
        }
    }
}
